crates/datasources/src/mysql/mod.rs

```diff
@@ -372,6 +374,62 @@ impl TableProvider for MysqlTableProvider {
             query,
             arrow_schema: projected_schema,
             metrics: ExecutionPlanMetricsSet::new(),
+            is_read_only: true,
+        }))
+    }
+
+    async fn insert_into(
+        &self,
+        state: &SessionState,
+        input: Arc<dyn ExecutionPlan>,
+        _overwrite: bool,
+    ) -> DatafusionResult<Arc<dyn ExecutionPlan>> {
+        let mut values = String::new();
+
+        let mut input = execute_stream(input, state.task_ctx())?;
+        while let Some(batch) = input.next().await {
+            let batch = batch?;
+
+            let mut row_sep = "";
+            for row_idx in 0..batch.num_rows() {
+                write!(&mut values, "{row_sep}(")?;
+                row_sep = ",\n\t";
+
+                let mut col_sep = "";
+                for col in batch.columns() {
+                    write!(&mut values, "{col_sep}")?;
+                    col_sep = ", ";
+
+                    let val = ScalarValue::try_from_array(col.as_ref(), row_idx)?;
+                    util::encode_literal_to_text(util::Datasource::MySql, &mut values, &val)
+                        .map_err(|e| DataFusionError::External(Box::new(e)))?;
+                }
+
+                write!(&mut values, ")")?;
+            }
+        }
+
+        if values.is_empty() {
+            let batch = create_count_record_batch(0);
+            let schema = batch.schema();
+            return Ok(Arc::new(MemoryExec::try_new(&[vec![batch]], schema, None)?));
+        }
+
+        let query = format!(
```

Comment:
this is sort of a rust question, but when do we do string formatting and when do we do string-building?

Reply:
formatting is just a more readable way imo

Reply:
What do you mean by "string-building"?

Reply:
for sure, though we're doing a lot of formatting here and above (sort of proceduraly building strings) so there's a lot of (probably? depends on how it gets optimized?) little string allocations, and a buffered string builder would be more efficient (and also frankly with the exception of this line probably easier to read.

Reply:
A buffered builder will definitely help (setting initial capacity a bit high, if that's what you're talking about). A rust `String` is made up of `Vec` so the capacity should double once it runs out.

Reply:
But that doesn't compete with `format!`. I think (and hope, never actually checked) it only allocates once, by calculating all the sizes and is more readable

Reply:
So true, but on line 496, we're using this `write!` macro call, and then writing that in this string, so we're presumably calling format a bunch of times.

Reply:
Oh, I see what you're asking. It's basically the same difference as pushing into a vector and allocating a new vector. `format!`, even though it would allocate once, would allocate all again and again if used in a loop. Building the string using `write!` doubles the allocated size of string during each, making it better for such cases.